content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+written as a logical formula - in particular, every such function has a CNF and
+a DNF.
+
+If we want to do that for other sets (apart from `B^n`, that is), we first have
+to be able to translate it into a set of boolean variables. As we will see,
+this can be easily done in the case we are interested in.
+
+All of this allows us to talk about what we mean by “computing a set is hard”.
+With this equivalency, checking if a given object is in a set becomes the
+*decision* problem of SAT - `x∈S`, exactly if `f[S](x)` evaluates to `true`.
+And checking if a set is empty becomes the *solution* problem of SAT - `S=∅`,
+exactly if `f[S]` is not satisfiable.
+
+With this, let us look at the specific sets we are interested in.
+
+## Basic interfaces as type sets
+
+Interfaces in Go have always described sets of types. For example, the interface
+
+```go
+type S interface {
+    X()
+    Y()
+    Z()
+}
+```
+
+is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.
+
+We also had a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
+
+
+```go
+type S interface { X() }
+type T interface { Y() }
+type U interface {
+    S
+    T
+}
+```
+
+This expresses `T=S∩T` as an interface. We can also view the predicate “has a
+method `X()`” as a boolean variable and think of this as the formula `X∧Y`.
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+```
+
+This meant that you could express interfaces which could never be satisfied by
+any type (i.e. which described an empty type set):
+
+```go
+interface{ X; NotX }
+```
+
+[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).
+
+So, we pretty much had a language to build boolean formulas, where every
+formula had the form
+
+    X∧Y∧¬Z
+
+Checking if a type implements an interface meant to check if it is in the set
+described by that interfaces. Which meant checking if the corresponding formula
+is satisfied.
+
+And checking if an interface is invalid meant checking if the set it describes
+is empty. Which meant checking if the formula is *satisfiable*. This was easy,
+though. If you look closely, those formulas are in DNF and as we noted above,
+solving SAT on a formula given in DNF is easy.
+
+## Adding unions
+
+Go 1.18 extends the interface syntax. For our purposes, the important addition
+is the `|` operator:
+
+```go
+type S interface{
+    A | B
+}
+```
+
+This represents the set of all types which are in the *union* of the type sets
+`A` and `B` - that is, the set of all types which are in `A` *or* in `B` (or
+both).
+
+This means our language of expressible formulas now also includes a
```

Comment:
I think this could expanding: why does `A | B` imply the `∨` operator? I think you need to say explicitly that `A | B` allows all operations provided by both `A` and `B`.

Reply:
Added an explanation. It's the opposite, BTW. "Allowed operations" are dual to "type sets", so the union of type sets allows all operations which are in *both* `A` and `B`.